File: src/model/Card.java
package model;

/**
 * Represents a playing card with a suit and rank.
 */
public class Card {
    private final Suit suit;
    private final Rank rank;

    public Card(Suit suit, Rank rank) {
        this.suit = suit;
        this.rank = rank;
    }

    public Suit getSuit() {
        return suit;
    }

    public Rank getRank() {
        return rank;
    }

    public int getValue() {
        return rank.getValue();
    }

    /**
     * Returns true if this card is a King (wild/joker).
     */
    public boolean isKing() {
        return rank.isKing();
    }

    /**
     * Returns true if this card is an Ace (can start a center pile).
     */
    public boolean isAce() {
        return rank.isAce();
    }

    /**
     * Returns true if this card is a Queen (completes a center pile).
     */
    public boolean isQueen() {
        return rank.isQueen();
    }

    /**
     * Returns true if this card is red (Hearts or Diamonds).
     */
    public boolean isRed() {
        return suit.isRed();
    }

    /**
     * Checks if this card can be placed on top of another card in a center pile.
     * King is wild and can be placed on any pile at any time (including to start).
     * 
     * @param topCard The current top card of the pile (null if pile is empty)
     * @return true if this card can be placed
     */
    public boolean canPlaceOn(Card topCard) {
        // King is wild - can be placed on any pile at any time
        if (this.isKing()) {
            return true; // King can start a pile or continue any pile
        }
        
        // If pile is empty, only Ace can start
        if (topCard == null) {
            return this.isAce();
        }
        
        // If top card is King (acting as wild), check if this card follows the expected value
        // The King takes the value of the card it replaced
        if (topCard.isKing()) {
            // We need to track what value the King represents
            // This will be handled by CenterPile
            return true;
        }
        
        // Normal case: this card's value must be exactly one more than top card
        return this.getValue() == topCard.getValue() + 1;
    }

    @Override
    public String toString() {
        return rank.getSymbol() + suit.getSymbol();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Card card = (Card) obj;
        return suit == card.suit && rank == card.rank;
    }

    @Override
    public int hashCode() {
        return 31 * suit.hashCode() + rank.hashCode();
    }
}

File: 21CardGame/mobile/src/models/Card.ts
import {
  Suit,
  Rank,
  getSuitSymbol,
  getRankSymbol,
  isSuitRed,
  isKing,
  isAce,
  isQueen,
} from './types';

/**
 * Represents a playing card with a suit and rank.
 */
export interface Card {
  readonly suit: Suit;
  readonly rank: Rank;
  readonly id: string; // Unique identifier for React keys
}

let cardIdCounter = 0;

/**
 * Create a new card
 */
export function createCard(suit: Suit, rank: Rank): Card {
  return {
    suit,
    rank,
    id: `card-${cardIdCounter++}`,
  };
}

/**
 * Get the numeric value of a card (based on rank)
 */
export function getCardValue(card: Card): number {
  return card.rank;
}

/**
 * Check if card is a King (wild/joker)
 */
export function isCardKing(card: Card): boolean {
  return isKing(card.rank);
}

/**
 * Check if card is an Ace (can start a center pile)
 */
export function isCardAce(card: Card): boolean {
  return isAce(card.rank);
}

/**
 * Check if card is a Queen (completes a center pile)
 */
export function isCardQueen(card: Card): boolean {
  return isQueen(card.rank);
}

/**
 * Check if card is red (Hearts or Diamonds)
 */
export function isCardRed(card: Card): boolean {
  return isSuitRed(card.suit);
}

/**
 * Check if this card can be placed on top of another card in a center pile.
 * King is wild and can be placed on any pile at any time (including to start).
 *
 * @param card The card to place
 * @param topCard The current top card of the pile (null if pile is empty)
 * @returns true if the card can be placed
 */
export function canCardPlaceOn(card: Card, topCard: Card | null): boolean {
  // King is wild - can be placed on any pile at any time
  if (isCardKing(card)) {
    return true;
  }

  // If pile is empty, only Ace can start
  if (topCard === null) {
    return isCardAce(card);
  }

  // If top card is King (acting as wild), we allow the next expected card
  // This will be handled by CenterPile which tracks expected value
  if (isCardKing(topCard)) {
    return true;
  }

  // Normal case: this card's value must be exactly one more than top card
  return getCardValue(card) === getCardValue(topCard) + 1;
}

/**
 * Get string representation of a card (e.g., "A♥", "K♠")
 */
export function cardToString(card: Card): string {
  return getRankSymbol(card.rank) + getSuitSymbol(card.suit);
}

/**
 * Check if two cards are equal
 */
export function cardsEqual(a: Card, b: Card): boolean {
  return a.suit === b.suit && a.rank === b.rank;
}

/**
 * Reset the card ID counter (useful for testing)
 */
export function resetCardIdCounter(): void {
  cardIdCounter = 0;
}
